org.eclipse.scout.rt.dataobject/src/main/java/org/eclipse/scout/rt/dataobject/id/IdCodec.java

```diff
@@ -99,19 +155,60 @@ public String toUnqualified(IId id) {
       if (mapper == null) {
         throw new PlatformException("Missing raw type mapper for wrapped type {}, id type {}", value.getClass(), id.getClass());
       }
-      return mapper.apply(value);
+      return encryptUnqualified(id.getClass(), mapper.apply(value), flags);
     }
     else if (id instanceof ICompositeId) {
       List<? extends IId> components = ((ICompositeId) id).unwrap();
-      return components.stream()
-          .map(this::toUnqualified)
+      Set<IIdCodecFlag> flagsWithoutEncryption = flags.stream()
+          .filter(Predicate.not(IdCodecFlag.ENCRYPTION::equals))
+          .collect(Collectors.toSet());
+      return encryptUnqualified(id.getClass(), components.stream()
+          .map(comp -> toUnqualified(comp, flagsWithoutEncryption))
           .map(s -> s == null ? "" : s) // empty string if component is null just in case of composite id
-          .collect(Collectors.joining(";"));
+          .collect(Collectors.joining(";")), flags);
     }
     else if (id instanceof UnknownId) {
-      return ((UnknownId) id).getId();
+      return encryptUnqualified(UnknownId.class, ((UnknownId) id).getId(), flags);
     }
-    return handleToUnqualifiedUnknownIdType(id);
+    return encryptUnqualified(id.getClass(), handleToUnqualifiedUnknownIdType(id, flags), flags);
+  }
+
+  /**
+   * Encrypts the given unqualifiedId iff
+   * <ul>
+   * <li>the given idClass needs to be encrypted (see {@link IdEncryption})
+   * <li>the given flags contain {@link IdCodecFlag#ENCRYPTION}
+   * </ul>
+   * This will create a signature (see {@link #createSignature(String)} and add it and a delimiter as a suffix to the
+   * given unqualifiedId.
+   */
+  protected String encryptUnqualified(Class<? extends IId> idClass, String unqualifiedId, Set<IIdCodecFlag> flags) {
+    if (isNullOrEmpty(unqualifiedId) || !isOneOf(IdCodecFlag.ENCRYPTION, flags) || !idInventory().isIdEncryption(idClass)) {
+      return unqualifiedId;
+    }
+    return unqualifiedId + ENCRYPTION_SIGNATURE_DELIMITER + createSignature(unqualifiedId);
+  }
+
+  /**
+   * Create a signature of the given unqualifiedId. Subclasses can override this method and use e.g. the current user id
+   * as well.
+   *
+   * @return an url safe signature
+   */
+  protected String createSignature(String unqualifiedId) {
+    if (isNullOrEmpty(unqualifiedId)) {
+      return unqualifiedId;
+    }
+    return encodeUrlSafe(createMac(getEncryptionPrivateKey(), unqualifiedId.getBytes(StandardCharsets.UTF_8)));
+  }
+
+  /**
+   * @return private key used to create a signature (see {@link #createSignature(String)}).
+   */
+  protected byte[] getEncryptionPrivateKey() {
+    return Optional.ofNullable(CONFIG.getPropertyValue(IdCrypterPasswordProperty.class))
```

Comment:
assert value and throw if no password set instead of calling createMac with null value (which will cause an error then)